src/main/java/com/snowplowanalytics/redash/RedashClient.java

```diff
@@ -0,0 +1,437 @@
+     * be set by value that was generated and returned by server in the response body.
+     * @throws IllegalArgumentException if there is already exists user-group with the name provided inside dataSource
+     * @throws IOException              if server is not available due to connection error or API key is invalid.
+     */
+    public int createUserGroup(UserGroup userGroup) throws IOException {
+        if (isEntityAlreadyExists(getUserGroups(), userGroup.getName())) {
+            throw new IllegalArgumentException(USER_GROUP_ALREADY_EXISTS);
+        }
+        String url = baseUrl + GROUPS_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        String returnValue = post(url, new Gson().toJson(userGroup), CheckResponseStatus.NO);
+        int id = getIdFromJson(returnValue);
+        userGroup.setId(id);
+        return id;
+    }
+
+    /**
+     * Adds user into user-group both specified by their ids.
+     *
+     * @return false if specified user-group already contains specified user, true if not and user was successfully added.
+     * @throws IllegalArgumentException if no user or no user group with provided userId and userGroupId accordingly exist
+     *                                  onto Redash server.
+     * @throws IOException              if server is not available due to connection error or API key is invalid.
+     */
+    public boolean addUserToGroup(int userId, int groupId) throws IOException {
+        checkIfEntityExists(User.class, userId);
+        UserGroup userGroup = getWithUsersAndDataSources(groupId);
+        if (userGroup.getUsers().stream().anyMatch(u -> u.getId() == userId)) {
+            return false;
+        }
+        String url = baseUrl + GROUPS_URL_PREFIX + "/" + groupId + MEMBERS_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        post(url, new JSONObject().put(USER_ID, userId).toString(), CheckResponseStatus.NO);
+        return true;
+    }
+
+    /**
+     * Adds data-source to user-group which are both specified by their ids in methods arguments.
+     *
+     * @return false if data-source already attached to group, true if not and has just added successfully to group
+     * @throws IllegalArgumentException if data-source or user-group does not exist
+     * @throws IOException              if server is not available due to connection error or API key is invalid.
+     */
+    public boolean addDataSourceToGroup(int dataSourceId, int groupId) throws IOException {
+        checkIfEntityExists(DataSource.class, dataSourceId);
+        UserGroup userGroup = getWithUsersAndDataSources(groupId);
+        if (userGroup.getDataSources().stream().anyMatch(ds -> ds.getId() == dataSourceId)) {
+            return false;
+        }
+        String url = baseUrl + GROUPS_URL_PREFIX + "/" + groupId + DATA_SOURCES_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        post(url, new JSONObject().put(DATA_SOURCE_ID, dataSourceId).toString(), CheckResponseStatus.NO);
+        return true;
```

Comment:
Same as above - how is the state of the server after this command being asserted?